### PR TITLE
Add the Helioprojective Radial Frame to coordinates

### PR DESCRIPTION
**Pull request by Cadair**
This is an addition to the coordinates package for the Helioprojective-Radial frame.

TODO:

- [x] Documentation. Specifcally the fact that the center of the disk is (0, -90) in this frame.

ping @LukeBarnard

### CODE REVIEW

File: sunpy/coordinates/transformations.py [hunk 1]
```diff
@@ -176,6 +176,57 @@ def hgs_to_hcc(heliogcoord, heliocframe):
...
+                                 HelioProjectiveRadial, HelioProjective)
+def hpr_to_hpc(hprframe, hpcframe):
+    """
+    Transform from the hprframe to a hpcframe
+    """
+    # Get params in rads
+    el = np.deg2rad(hprframe.psi)
+    pa = np.deg2rad(hprframe.dec)
+    # pi/2 correction
+    el += u.deg*np.pi/2
```
Comment (dpshelio) [inline sunpy/coordinates/transformations.py]: what? pi/2 * u.deg ??? is that right?

Comment (Cadair) [inline sunpy/coordinates/transformations.py]: probably not, this doesn't work.

Comment (dpshelio) [inline sunpy/coordinates/transformations.py]: maybe if you use `u.rad` instead ;)

File: sunpy/coordinates/transformations.py [hunk 2]
```diff
@@ -176,6 +176,57 @@ def hgs_to_hcc(heliogcoord, heliocframe):
...
+    Transform from the hpcframe to a hprframe
+    """
+    # Put it in rads for np
+    lon = np.deg2rad(hpcframe.Tx)
+    lat = np.deg2rad(hpcframe.Ty)
+    # Elongation calc:
+    # Get numerator and denomenator for atan2 calculation
+    top = np.cos(lat)*np.cos(lon)
+    btm = np.sqrt((np.cos(lat)**2)*(np.sin(lon)**2) + (np.sin(lat)**2))
+    el = np.arctan2(top,btm)
```
Comment (LukeBarnard) [inline sunpy/coordinates/transformations.py]: Looking at Eq19 from Thompson, and his notation that arg(x,y) = arctan(y/x), I think the top and bottom need swapping round in this call to arctan2. Same applies to the pa calculation on Line 196. However, when I tried this it returned ValueError: Latitude angle(s) must be within -90 deg <= angle <= 90 deg

File: sunpy/coordinates/transformations.py [hunk 3]
```diff
@@ -20,7 +20,7 @@
 from .representation import (SphericalWrap180Representation,
                              UnitSphericalWrap180Representation)
 from .frames import (HeliographicStonyhurst, HeliographicCarrington,
-                     Heliocentric, Helioprojective)
+                     Heliocentric, Helioprojective, HelioprojectiveRadial)
```
Comment (LukeBarnard) [inline sunpy/coordinates/transformations.py]: I think either this needs to be HelioProjectiveRadial, or the class name needs to change in frames.py

**Comment by Cadair** [comment]
@LukeBarnard I think I have this working now. I don't know if you want to take it for a spin?

**Comment by LukeBarnard** [comment]
Will do! Won't have time before the festive shutdown, but will have a crack at it early January.

**Comment by Cadair** [comment]
Some variant on this maybe: https://github.com/astropy/astropy/blob/v2.0.x/astropy/coordinates/representation.py#L1661

**Comment by Cadair** [comment]
Ok, this got hard and complex, *however* some progress is being made.

This PR now needs [this](https://github.com/astropy/astropy/compare/master...Cadair:reversePhysicsSpherical) astropy branch to properly function.

I am also planning on moving the representations added here into astropy, but they are here for the moment.

**Comment by elastufka** [comment]
Hi, has there been any progress on this lately? It would be great to have this feature, if only for on-disk coordinates.

**Comment by ayshih** [comment]
No progress, largely due to lack of interest.  Separate from this PR, I drafted an implementation as well (https://github.com/ayshih/sunpy/commit/dd3b1f0ef58709e50c7b4674d817cf01b0730b0e), and my version has pluses and minuses compared to this version.  Since you're actually interested in having this functionality, can you create a proper issue requesting it?

**Comment by svank** [comment]
Hi @Cadair @ayshih et al, I'm interested in this feature, so I'm planning to pick things up from here and see if I can make any progress on the implementation. I know it's been a while and the landscape has changed, but if there are any suggestions, warnings, etc. that you recall, I'd appreciate hearing them.

**Comment by Cadair** [comment]
This got really hard once I tried to make spherical representations which put 0 deg latitude at the south pole of the coordinate frame. I would be in favour of implementing a coordinate system where we have -90 at the disk center and then taking it from there.

**Comment by svank** [comment]
I've been playing with this at https://github.com/svank/sunpy/tree/hpr and as far as I can tell, the implementation in this PR works very well at what the `astropy.coordinates` module does (though I won't pretend to have a deep understanding of the module and all it can do). With the SouthPoleSphericalRepresentation, it seems to do a great job at representing HPR coordinates properly (with the Sun/SP at 0 latitude) and converting to/from HPC coords, and I think the only broken things I found were just bitrot. Do you recall any major TODOs here?

What I've noticed as missing is full interoperability with `astropy.wcs`, where the 90deg offset between true HPR and FITS-encoded HPR isn't handled correctly in functions like `pixel_to_world`, and support for applying that offset in `WCSAxes`. (I suppose I don't even know if these things were a goal here, but they are where I'm most interested in using HPR coordinates). `WCSAxes` has hardcoded special cases for some frames to decide whether longitude should wrap at 180 or 360, so maybe it could also special-case HPR to add 90 to the FITS file latitudes (though that must be a separate astropy PR). For interacting with `wcs`, is that what the `FITSHelioprojectiveRadial` frame in this PR was meant to end up doing? Maybe `hpr_wcs.pixel_to_world` produces a `FITSHelioprojectiveRadial` coordinate instead of `HelioprojectiveRadial`, and `FITSHelioprojectiveRadial` knows how to handle HPR coordinates as embedded in the FITS file?

If that's sort of the direction you were thinking way back when, that seems like an approach that keeps in all in sunpy. Any reason for preferring that over trying to have `astropy.wcs` handle the offset between true HPR and FITS-encoded HPR? It's maybe not the most astropy-specific special case, but applying the offset there seems like it would make for easy interoperability with the `coordinates` implementation here as well as direct use of proper HPR coordinates in `wcs` and anything else that builds on it.

**Pull request closed without merging**